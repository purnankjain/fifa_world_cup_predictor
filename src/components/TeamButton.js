import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './../styles/css/teamButton.css';

export default class TeamButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name : props.name,
      winner : false
    }
    this.setWinner = this.setWinner.bind(this);
  }

  setWinner(e) {
    this.setState({winner : !this.state.winner});
  }

  render() {
    var styleClass = this.state.winner ? 'winner' : '';
    return (<div onClick={this.setWinner} className={styleClass}>{this.state.name}</div>);
  }
}

TeamButton.propTypes = {
  name: PropTypes.string.isRequired
};
