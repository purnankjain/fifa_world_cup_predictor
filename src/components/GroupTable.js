import React, { Component } from 'react';
import TeamButton from './TeamButton';
import PropTypes from 'prop-types';

export default class GroupTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      groupWinner : ''
    };
    this.setGroupWinner = this.setGroupWinner.bind(this);
    this.rollbackGroupWinner = this.rollbackGroupWinner.bind(this);
  }
  setGroupWinner(e) {
    this.setState({ groupWinner: e.target.innerHTML });
  }
  rollbackGroupWinner() {
    this.setState({ groupWinner: '' });
  }
  isTeamWinner(teamName) {
    return teamName === this.state.groupWinner;
  }
  render() {
    var teamButtons = this.props.teams.map((team, index) => <TeamButton name={team} key={index} winnerCallback = {this.setGroupWinner} rollbackWinnerCallback={this.rollbackGroupWinner} isWinner={this.isTeamWinner(team)}/>);
    return (<div>
      <h3>{this.props.name}</h3>
      {teamButtons}
      </div>)
  }
}

GroupTable.propTypes = {
  name : PropTypes.string.isRequired,
  teams : PropTypes.arrayOf(PropTypes.string).isRequired
}
