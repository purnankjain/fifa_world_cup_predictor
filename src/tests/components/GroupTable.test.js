import React from 'react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import TestUtils from 'react-dom/test-utils';
import GroupTable from '../../components/GroupTable'
import TeamButton from '../../components/TeamButton';

var teams = ['India', 'Pakistan', 'Sri Lanka'];
var groupName = "Group A";
var groupTableComponent = <GroupTable name={groupName} teams={teams}/>;
var _WINNER_CLASS_NAME = 'winner';

describe('GroupTable component test', () => {
  it('should match snapshot', () => {
    var rendered = renderer.create(groupTableComponent).toJSON();

    expect(rendered).toMatchSnapshot();
  });

  it('should render multiple teamButtons', () => {
    var rendered = renderer.create(groupTableComponent).root;

    expect(rendered.findAllByType(TeamButton).length).toEqual(teams.length);
  });

  it('should set member team button as winner on click', () => {
    var group = TestUtils.renderIntoDocument(groupTableComponent);
    var groupNode = ReactDOM.findDOMNode(group);

    TestUtils.Simulate.click(groupNode.childNodes[1]);

    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME).length).toEqual(1);
    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME)[0].innerHTML).toEqual(teams[0]);
  });

  it('should rollback winner on click', () => {
    var group = TestUtils.renderIntoDocument(groupTableComponent);
    var groupNode = ReactDOM.findDOMNode(group);

    TestUtils.Simulate.click(groupNode.childNodes[1]);
    TestUtils.Simulate.click(groupNode.childNodes[1]);

    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME).length).toEqual(0);
  });

  it('should have only one winner', () => {
    var group = TestUtils.renderIntoDocument(groupTableComponent);
    var groupNode = ReactDOM.findDOMNode(group);

    TestUtils.Simulate.click(groupNode.childNodes[1]);
    TestUtils.Simulate.click(groupNode.childNodes[2]);

    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME).length).toEqual(1);
    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME)[0].innerHTML).toEqual(teams[0]);
  });

  it('should have 2 winners when groupTable is made for 2 winners', () => {
    var winnerCount = 2;
    var groupTableComponent = <GroupTable name={groupName} teams={teams} winnerCount={winnerCount}/>;
    var group = TestUtils.renderIntoDocument(groupTableComponent);
    var groupNode = ReactDOM.findDOMNode(group);

    TestUtils.Simulate.click(groupNode.childNodes[1]);
    TestUtils.Simulate.click(groupNode.childNodes[2]);
    TestUtils.Simulate.click(groupNode.childNodes[3]);

    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME).length).toEqual(2);
    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME)[0].innerHTML).toEqual(teams[0]);
    expect(groupNode.getElementsByClassName(_WINNER_CLASS_NAME)[1].innerHTML).toEqual(teams[1]);
  });

  it('should make a callback on winner update', () => {
    var winnerUpdateCallbackCount = 0;
    var winnerUpdateCallback = () => { winnerUpdateCallbackCount++ };
    var groupTableComponent = <GroupTable name={groupName} teams={teams} winnerUpdateCallback={winnerUpdateCallback}/>;
    var group = TestUtils.renderIntoDocument(groupTableComponent);
    var groupNode = ReactDOM.findDOMNode(group);

    TestUtils.Simulate.click(groupNode.childNodes[1]);

    expect(winnerUpdateCallbackCount).toEqual(1);
  });
});
