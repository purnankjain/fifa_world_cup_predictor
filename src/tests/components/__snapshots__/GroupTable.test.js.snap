// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupTable component test should match snapshot 1`] = `
<div>
  <h3>
    Group A
  </h3>
  <div
    className=""
    onClick={[Function]}
  >
    India
  </div>
  <div
    className=""
    onClick={[Function]}
  >
    Pakistan
  </div>
  <div
    className=""
    onClick={[Function]}
  >
    Sri Lanka
  </div>
</div>
`;
